File: Haskell/BubbleSort.hs
-- Source: ChatGPT

bubbleSort :: Ord a => [a] -> [a]
bubbleSort [] = []
bubbleSort xs = bubbleSort' xs (length xs)

bubbleSort' :: Ord a => [a] -> Int -> [a]
bubbleSort' xs 0 = xs
bubbleSort' xs n = bubbleSort' (bubble xs) (n-1)

bubble :: Ord a => [a] -> [a]
bubble [] = []
bubble [x] = [x]
bubble (x:y:xs) 
    | x > y     = y : bubble (x:xs)
    | otherwise = x : bubble (y:xs)


main :: IO ()
main = do
  let unsortedList = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]
      sortedList = bubbleSort unsortedList
  putStrLn "Length of Bubble Sorted List:"
  print (length (sortedList))

File: Languages/Prolog/BubbleSort.pl
:- style_check(-singleton).
% Source: ChatGPT
:- dynamic logging/1.

:- assert(logging(0)).

bubble_sort(List, Sorted) :-
    length(List, Length),
    length(Sorted, Length),
    bubble_sort_(List, Sorted).

bubble_sort_([A,B|Rest], [B,A|SortedRest]) :-
    A > B,
    !,
    bubble_sort_(Rest, SortedRest).

bubble_sort_([Head|Rest], [Head|SortedRest]) :-
    bubble_sort_(Rest, SortedRest).

bubble_sort_([], []) :- !.

main :-
    List = [10650,56341,77676,47158,34999,22009,19680,41870,31090,30668],
    bubble_sort(List, SortedList),
    (logging(1) ->
        write('Sorted list: '),
        write(SortedList), nl
    ;logging(0) -> 
        nl).


